libs/libc/stdio/lib_fopencookie.c

```diff
@@ -0,0 +1,181 @@
+#include <stdlib.h>
+#include <unistd.h>
+#include <fcntl.h>
+#include <string.h>
+#include <assert.h>
+#include <errno.h>
+#include <nuttx/tls.h>
+
+#include "libc.h"
+
+/****************************************************************************
+ * Private Types
+ ****************************************************************************/
+
+static ssize_t freadcookie(FAR void *s, FAR char *buf, size_t nbytes);
+static ssize_t fwritecookie(FAR void *s, FAR const char *buf, size_t nbytes);
+static off_t fseekcookie(FAR void *s, off_t offset, int whence);
+static int fclosecookie(FAR void *s);
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+static FAR FILE *fopencookie_init_stream(int oflags)
+{
+  FAR FILE *stream;
+
+  stream = lib_zalloc(sizeof(FILE));
+  if (stream == NULL)
+    {
+      set_errno(-ENOMEM);
+      return NULL;
+    }
+
+  nxrmutex_init(&stream->fs_lock);
+
+#ifndef CONFIG_STDIO_DISABLE_BUFFERING
+#if CONFIG_STDIO_BUFFER_SIZE > 0
+  /* Set up pointers */
+
+  stream->fs_bufstart = stream->fs_buffer;
+  stream->fs_bufend   = &stream->fs_bufstart[CONFIG_STDIO_BUFFER_SIZE];
+  stream->fs_bufpos   = stream->fs_bufstart;
+  stream->fs_bufread  = stream->fs_bufstart;
+  stream->fs_flags    = __FS_FLAG_UBF; /* Fake setvbuf and fclose */
+
+#ifdef CONFIG_STDIO_LINEBUFFER
+  /* Setup buffer flags */
+
+  stream->fs_flags   |= __FS_FLAG_LBF; /* Line buffering */
+
+#endif /* CONFIG_STDIO_LINEBUFFER */
+#endif /* CONFIG_STDIO_BUFFER_SIZE > 0 */
+#endif /* CONFIG_STDIO_DISABLE_BUFFERING */
+
+  /* Save the file description and open flags.  Setting the
+   * file descriptor locks this stream.
+   */
+
+  stream->fs_fd       = -1;
+  stream->fs_oflags   = oflags;
+
+  return stream;
+}
+
+static ssize_t freadcookie(FAR void *s, FAR char *buf, size_t nbytes)
+{
+  FAR FILE *stream = (FAR FILE *)s;
+  FAR struct cookie_s *cookie = (FAR struct cookie_s *)&stream->fs_cookie;
+  int ret;
+
+  DEBUGASSERT(cookie->cookie != NULL && cookie->cookie_io.read != NULL);
+
+  ret = cookie->cookie_io.read(cookie->cookie, buf, nbytes);
+  return ret;
+}
+
+static ssize_t fwritecookie(FAR void *s, FAR const char *buf, size_t nbytes)
+{
+  FAR FILE *stream = (FAR FILE *)s;
+  FAR struct cookie_s *cookie = (FAR struct cookie_s *)&stream->fs_cookie;
+  int ret;
+
+  DEBUGASSERT(cookie->cookie != NULL && cookie->cookie_io.write != NULL);
+
+  ret = cookie->cookie_io.write(cookie->cookie, buf, nbytes);
+  return ret;
+}
+
+static off_t fseekcookie(FAR void *s, off_t offset, int whence)
+{
+  FAR FILE *stream = (FAR FILE *)s;
+  FAR struct cookie_s *cookie = (FAR struct cookie_s *)&stream->fs_cookie;
+  int ret;
+
+  DEBUGASSERT(cookie->cookie != NULL && cookie->cookie_io.seek != NULL);
+
+  ret = cookie->cookie_io.seek(cookie->cookie, &offset, whence);
+  return ret;
+}
+
+static int fclosecookie(FAR void *s)
```

Comment:
let's remove f*cookie and forward to cookie_io_functions_t directly

Reply:
Done